yggdrasil_decision_forests/learner/decision_tree/oblique.cc

```diff
@@ -788,12 +788,31 @@ void SampleProjection(const absl::Span<const int>& features,
     }
   };
 
+  #ifndef NDEBUG  // Keep DCHECK_EQ from for feature : features
   for (const auto feature : features) {
     DCHECK_EQ(data_spec.columns(feature).type(), dataset::proto::NUMERICAL);
-    if (unif01(*random) < projection_density) {
-      projection->push_back({feature, gen_weight(feature)});
-    }
   }
+  #endif
+
+  const size_t p = features.size();
+  std::binomial_distribution<size_t> binom(p, projection_density);
+
+  // Exp[Binomial(p,projection_density)] = k
+  const size_t k = binom(*random);
+
+  absl::btree_set<size_t> picked_idx;
```

Comment:
Open question:
What would be the speed impact of using a std::vector(bool) or vector(u8)?

btree_set has a good theoretical space complexity, but effectively creates multiple heap allocations. vector<bool> would be a larger but single heap allocation.

Bonus (not for this PR): The heap allocation of a vector<bool> will be easier to cache and reuse over multiple calls of this method.

Reply:
*The YDF team will handle it after this PR is submited